libs/ledger/include/ledger/chain/transaction_encoding.hpp

```diff
@@ -0,0 +1,206 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/byte_array/byte_array.hpp"
+#include "core/serializers/byte_array_buffer.hpp"
+#include "meta/type_traits.hpp"
+#include "vectorise/platform.hpp"
+
+namespace fetch {
+namespace ledger {
+namespace detail {
+
+template <typename T>
+meta::IfIsUnsignedInteger<T, uint64_t> ToU64(T value)
+{
+  return value;
+}
+
+template <typename T>
+meta::IfIsSignedInteger<T, uint64_t> ToU64(T value)
+{
+  return static_cast<uint64_t>(std::abs(value));
+}
+
+template <typename T>
+meta::IfIsInteger<T, fetch::byte_array::ConstByteArray> EncodeInteger(T value)
+{
+  using fetch::byte_array::ByteArray;
+
+  bool const is_signed = meta::IsSignedInteger<T> && (value < 0);
+
+  ByteArray encoded;
+  if (!is_signed && (value <= T{0x7f}))
+  {
+    encoded.Resize(1);
+    encoded[0] = static_cast<uint8_t>(value & 0x7f);
+  }
+  else
+  {
+    uint64_t const abs_value = ToU64(value);
+
+    if (is_signed && (abs_value <= 0x1F))
```

Comment:
Is signed check should not be necessary, given the `else` context.

Reply:
Nope, still necessary because it might be unsigned but greater that 7bits in length